### ISSUE TITLE
Operators dont get their distributor information from the current grid

### ISSUE THREAD

**Issue by mikeando** [opesci/devito#998]
At the moment it seems that the MPI information used by the operator are baked into it when it is constructed, rather than taken from its arguments.

There are checks in place that all the arguments are on the same grid - but this may have a different distributor than the operator was built with. 

Currently I can work around this in my examles by doing:

~~~
    m={}
    m['comm'] = u.grid.distributor._obj_comm
    m['nb'] = u.grid.distributor._obj_neighborhood
    op.objects = tuple( m.get(o.name, o) for o in op.objects)
    op.apply(u=u)
~~~

but it feels pretty hacky.

A complete example can be found at 

https://gist.github.com/mikeando/35f661d13b374e857315859cd0d975d6

**Comment by mikeando** [comment]
I tried fixing this by patching operator.py to contain something like

~~~
        # Process Objects (which may need some `args`)
        for o in self.objects:
            args.update(o._arg_values(args, **kwargs))
+
+    if grid is not None:
+        args.update(grid._arg_defaults())
~~~

but was unable to get that to work.

**Comment by mikeando** [comment]
For the MPI + cloudpickle version in that example to run devito/mpi/distributed.py must be patched to create SELF based distribtors rather than WORLD by default. That is done using the following patch.

~~~
diff --git a/devito/mpi/distributed.py b/devito/mpi/distributed.py
index 8801da2f..e607f055 100644
--- a/devito/mpi/distributed.py
+++ b/devito/mpi/distributed.py
@@ -181,7 +181,7 @@ class Distributor(AbstractDistributor):
             if not MPI.Is_initialized():
                 MPI.Init()
 
-            self._input_comm = (input_comm or MPI.COMM_WORLD).Clone()
+            self._input_comm = (input_comm or MPI.COMM_SELF).Clone()
 
             # Make sure the cloned communicator will be freed up upon exit
             def cleanup():
~~~~

**Comment by rhodrin** [comment]
@mikeando,

Thanks for opening this issue. It will be looked into soon.

**Comment by rhodrin** [comment]
@mikeando Just a quick update. I've been looking into this today and the problem is probably to do with `Operator` caching some things that it shouldn't resulting in `comm` and `nb` not being updated 'automatically' as they should be. Pretty sure I know roughly the area where some changes are needed to fix this, just need to figure out the best way of doing so.

**Comment by rhodrin** [comment]
@mikeando When you run the example code you pasted above what run times are you seeing for each of the Operators?
I'm seeing an issue, where if I 'fix' the required argument values (through some means) I get extremely slow operator executions. So for example, for the code you pasted, running on 4 ranks, the first two operators take <1s but the final 'fixed distributed' operator takes > 100s.

**Comment by mikeando** [comment]
I get about 0.17s to 0.33s for the local operator, 
About 0.7s for the broken distributed operator
And the "fixed" one takes 378s. 

So, something is still definitely fishy. I'd expect a little slow down using MPI, but a 1000x time slow down is probably more than expected.

**Comment by rhodrin** [comment]
Hi @mikeando,

Could you check if the issues you're seeing are fixed on https://github.com/opesci/devito/tree/rebuild-op.objects when you get the chance please?

Also, I'm not entirely sure what is causing the 'hanging'/long operator run times in the code you posted, but it seems to be related to writing the output files. For example, I'm not seeing any hanging with the following example:
```
import numpy as np
import cloudpickle

from mpi4py import MPI

from devito import Function, TimeFunction, Grid, Eq, Operator, configuration
from devito import norm

configuration["mpi"] = True
configuration["openmp"] = True

def build_operator():
    toy_grid = Grid(shape=(100,100), comm=MPI.COMM_SELF)
    u = TimeFunction(name='u',grid=toy_grid, space_order=4, time_order=1)
    eqn = Eq(u.forward, u + 0.001 + 0.001 * u.dx2)
    op = Operator(eqn)
    return op

def get_operator(comm):
    op = None
    if comm is None or comm.rank == 0:
        op = build_operator()

    if comm is not None:
        if comm.rank == 0:
            op_pickle = cloudpickle.dumps(op)
        else:
            op_pickle = None
        op_pickle = comm.bcast(op_pickle, root=0)
        if comm.rank != 0:
            op = cloudpickle.loads(op_pickle)
        del op_pickle

    assert op is not None

    return op

op = get_operator(None)

def setup_u(g:Grid) -> TimeFunction:
    u = TimeFunction(name="u", grid=g, space_order=4, time_order=1)
    return u


def run_op(op, u):
    op.apply(u=u, time_M=1)
    print('norm = ', norm(u))

def run_local(op):
    g = Grid(shape=(100,100), comm=MPI.COMM_SELF)
    u = setup_u(g)
    run_op(op, u)

def run_dist(op):
    g = Grid(shape=(100,100), comm=MPI.COMM_WORLD)
    u = setup_u(g)
    run_op(op, u)

run_local(op)
run_dist(op)
```

**Comment by rhodrin** [comment]
So I'm still seeing this 'hanging' in certain examples when I wouldn't expect it. Still looking into that.

**Comment by rhodrin** [comment]
Could you also confirm that removing `configuration["openmp"] = True` results in operators having more sensible run times in all cases?

**Comment by rhodrin** [comment]
>I get about 0.17s to 0.33s for the local operator,
About 0.7s for the broken distributed operator
And the "fixed" one takes 378s.
>
>So, something is still definitely fishy. I'd expect a little slow down using MPI, but a 1000x time slow down is probably more than expected.

So I was just being silly - these 'slowdowns' were for me coming from me not obeying `# MPI_RANKS * # OMP_THREADS_PER_RANK <= #PHYSICAL_CORES`.
So I think the patch on the branch posted above should rectify this issue. It will be tidied and PR'd soon.

**Comment by rhodrin** [comment]
Fixed in #1016.
